stage0/src/dice_attestation.rs

```diff
@@ -0,0 +1,163 @@
+    let hkdf = Hkdf::<Sha256>::new(Some(ID_SALT.as_bytes()), ikm);
+    let mut okm: [u8; SIGNATURE_LENGTH] = [0u8; SIGNATURE_LENGTH];
+    hkdf.expand(info.unwrap_or(&[]), &mut okm)
+        .expect("20 is a valid length for Sha256 to output");
+
+    okm
+}
+
+fn generate_ecdsa_keys(info: &str) -> (SigningKey, [u8; SIGNATURE_LENGTH]) {
+    let key = SigningKey::random(&mut OsRng);
+    let public_key = VerifyingKey::from(&key);
+    let key_id = derive_id(
+        public_key.to_encoded_point(false).as_bytes(),
+        Some(info.as_bytes()),
+    );
+    (key, key_id)
+}
+
+fn generate_stage1_ca_cwt(
+    stage0_eca_key: SigningKey,
+    stage0_cert_id_hex: String,
+    measurements: &Measurements,
+) -> Result<Vec<u8>, CoseError> {
+    // Generate Stage 1 keys and Signer.
+    let info_str = "ID";
+    let stage1_ca_key = generate_ecdsa_keys(info_str);
+    let stage1_ca_verifying_key = VerifyingKey::from(&stage1_ca_key.0);
+    let stage0_signer = Stage0Signer {
+        signing_key: stage0_eca_key,
+    };
+
+    // Generate claims and add the Stage1 CA key.
+    let claims = cwt::ClaimsSetBuilder::new()
+        .issuer(stage0_cert_id_hex)
+        .subject(hex::encode(stage1_ca_key.1))
+        .cwt_id(Vec::from(stage1_ca_key.1))
+        // Add additional private-use claim.
+        .private_claim(
+            -4670552,
+            Value::Bytes(Vec::from(
+                stage1_ca_verifying_key.to_encoded_point(false).as_bytes(),
+            )),
+        )
+        .private_claim(
+            -4670555,
+            Value::Bytes(Vec::from(measurements.kernel_measurement)),
+        )
+        .private_claim(
+            -4670556,
+            Value::Bytes(Vec::from(measurements.kernel_measurement)),
+        )
+        .private_claim(
+            -4670557,
+            Value::Bytes(Vec::from(measurements.setup_data_measurement)),
+        )
+        .private_claim(
+            -4670558,
+            Value::Bytes(Vec::from(measurements.ram_disk_measurement)),
+        )
+        .private_claim(
+            -4670559,
+            Value::Bytes(Vec::from(measurements.memory_map_measurement)),
+        )
+        .private_claim(
+            -4670560,
+            Value::Bytes(Vec::from(measurements.acpi_measurement)),
+        )
+        .build();
+
+    let aad = b"";
+
+    // Build a `CoseSign1` object.
+    let protected = coset::HeaderBuilder::new()
+        .algorithm(iana::Algorithm::ES384)
+        .build();
+    let unprotected = coset::HeaderBuilder::new()
+        .key_id((*b"AsymmetricECDSA384").into())
+        .build();
+    let sign1 = coset::CoseSign1Builder::new()
+        .protected(protected)
+        .unprotected(unprotected)
+        .payload(claims.clone().to_vec()?)
+        .create_signature(aad, |data| stage0_signer.sign(data))
+        .build();
+    sign1.to_vec()
+}
+
+pub fn generate_stage1_attestation(measurements: &Measurements) {
```

Comment:
This function does more than generating attestation (in the doc we call it `CertificateAuthority`), i.e. it also sends it to the Stage1